File: qbtools/commands/tagging.py
import tldextract
import collections

from qbtools import utils
from datetime import datetime
from qbittorrentapi import TrackerStatus


DEFAULT_TAGS = [
    "activity:",
    "added:",
    "dupe",
    "expired",
    "not-linked",
    "not-working",
    "unregistered",
    "tracker-down",
    "site:",
]

UNREGISTERED_MATCHES = [
    "UNREGISTERED",
    "TORRENT NOT FOUND",
    "TORRENT IS NOT",
    "NOT REGISTERED",
    "NOT EXIST",
    "UNKNOWN TORRENT",
    "TRUMP",
    "RETITLED",
    "INFOHASH NOT FOUND",  # blutopia
    "TORRENT HAS BEEN DELETED",  # blutopia
    "DEAD",
    "DUPE",
    "COMPLETE SEASON UPLOADED",
    "PROBLEM",
    "SPECIFICALLY BANNED",
    "OTHER",
    "NUKED",
    "INVALID INFOHASH",
]

MAINTENANCE_MATCHES = [
    "DOWN",
    "UNREACHABLE",
    "BAD GATEWAY",
    "TRACKER UNAVILABLE",
]


def __init__(app, logger):
    logger.info("Tagging torrents in qBittorrent...")

    today = datetime.today()
    torrents = app.client.torrents.info()
    config = app.config.get("trackers", [])
    config = {y: x for x in config for y in x["urls"]}

    exclude_categories = [i for s in app.exclude_category for i in s]
    if exclude_categories:
        torrents = list(
            filter(lambda x: x.category not in exclude_categories, torrents)
        )

    exclude_tags = [i for s in app.exclude_tag for i in s]
    if exclude_tags:
        torrents = list(
            filter(lambda x: any(y not in x.tags for y in exclude_tags), torrents)
        )

    extractTLD = tldextract.TLDExtract(cache_dir=None)
    tags = collections.defaultdict(list)
    paths = []

    for t in torrents:
        tags_to_add = []
        trackers = list(filter(lambda s: s.tier >= 0, t.trackers))  # Expensive

        url = t.tracker
        if not url and trackers:
            url = trackers[0].url

        tracker = config.get(extractTLD(url).registered_domain)

        if app.added_on:
            tags_to_add.append(calculate_date_tags("added", t.added_on, today))

        if app.last_activity:
            tags_to_add.append(calculate_date_tags("activity", t.last_activity, today))

        if app.sites:
            if tracker:
                tags_to_add.append(f"site:{tracker['name']}")
            else:
                tags_to_add.append(f"site:unmapped")

        if app.unregistered or app.tracker_down or app.not_working:
            if not any(s.status is TrackerStatus.WORKING for s in trackers):
                messages = [z.msg.upper() for z in trackers]
                if app.unregistered and any(
                    match in msg for msg in messages for match in UNREGISTERED_MATCHES
                ):
                    tags_to_add.append("unregistered")
                elif app.tracker_down and any(
                    match in msg for msg in messages for match in MAINTENANCE_MATCHES
                ):
                    tags_to_add.append("tracker-down")
                elif app.not_working:
                    tags_to_add.append("not-working")

        if app.expired and tracker and t.state_enum.is_complete:
            if (
                tracker["required_seed_ratio"]
                and t.ratio >= tracker["required_seed_ratio"]
            ):
                tags_to_add.append("expired")
            elif tracker["required_seed_days"] and t.seeding_time >= utils.seconds(
                tracker["required_seed_days"]
            ):
                tags_to_add.append("expired")

        if app.duplicates:
            if t.content_path in paths and not t.content_path == t.save_path:
                tags_to_add.append("dupe")
            else:
                paths.append(t.content_path)

        if app.not_linked and not utils.is_linked(t.content_path):
            tags_to_add.append("not-linked")

        for tag in tags_to_add:
            tags[tag].append(t)

    for tag, tagged in sorted(tags.items()):
        old_hashes = [t.hash for t in torrents if tag in t.tags and not t in tagged]
        new_hashes = [t.hash for t in tagged if not tag in t.tags]

        if old_hashes:
            app.client.torrents_remove_tags(tags=tag, torrent_hashes=old_hashes)

        if new_hashes:
            app.client.torrents_add_tags(tags=tag, torrent_hashes=new_hashes)

        if old_hashes or new_hashes:
            logger.info(
                f"{tag} - untagged {len(old_hashes)} old and tagged {len(new_hashes)} new"
            )

    empty_tags = list(
        filter(
            lambda tag: tag not in tags
            and any(tag.lower().startswith(x.lower()) for x in DEFAULT_TAGS),
            app.client.torrents_tags(),
        )
    )
    if empty_tags:
        app.client.torrents_delete_tags(tags=empty_tags)
        logger.info(f"Removed {len(empty_tags)} old tags from qBittorrent")

    logger.info("Finished tagging torrents in qBittorrent")


def calculate_date_tags(prefix, timestamp, today):
    diff = today - datetime.fromtimestamp(timestamp)
    if diff.days == 0:
        return f"{prefix}:1d"
    elif diff.days <= 7:
        return f"{prefix}:7d"
    elif diff.days <= 30:
        return f"{prefix}:30d"
    elif diff.days <= 180:
        return f"{prefix}:180d"
    else:
        return f"{prefix}:>180d"


def add_arguments(command, subparser):
    """
    Description:
        Tag torrents. This command can be used to tag torrents with various tags, such as torrents that have not been active for a while, torrents that have not been working for a while, torrents that have expired an ratio or seeding time, torrents that have the same content path, etc.
    Usage:
        qbtools.py tagging --help
    Example:
        # Tag torrents
        qbtools.py tagging --exclude-category manual --added-on --expired --last-activity --sites --unregistered
    """
    parser = subparser.add_parser(command)
    parser.add_argument(
        "--exclude-category",
        nargs="*",
        action="append",
        metavar="mycategory",
        default=[],
        help="Exclude all torrent with this category, can be repeated multiple times",
        required=False,
    )
    parser.add_argument(
        "--exclude-tag",
        nargs="*",
        action="append",
        metavar="mytag",
        default=[],
        help="Exclude all torrents with this tag, can be repeated multiple times",
        required=False,
    )
    parser.add_argument(
        "--added-on",
        action="store_true",
        help="Tag torrents with added date (last 24h, 7 days, 30 days, etc)",
    )
    parser.add_argument(
        "--duplicates",
        action="store_true",
        help="Tag torrents with the same content path",
    )
    parser.add_argument(
        "--expired",
        action="store_true",
        help="Tag torrents that have an expired ratio or seeding time (defined in config.yaml)",
    )
    parser.add_argument(
        "--last-activity",
        action="store_true",
        help="Tag torrents with last activity date (last 1d, 7 days, 30 days, etc)",
    )
    parser.add_argument(
        "--not-linked",
        action="store_true",
        help="Tag torrents with files without hardlinks or symlinks, use with filtering by category/tag",
    )
    parser.add_argument(
        "--not-working",
        action="store_true",
        help="Tag torrents with not working tracker status",
    )
    parser.add_argument(
        "--sites",
        action="store_true",
        help="Tag torrents with site names (defined in config.yaml)",
    )
    parser.add_argument(
        "--tracker-down",
        action="store_true",
        help="Tag torrents with temporarily down trackers",
    )
    parser.add_argument(
        "--unregistered",
        action="store_true",
        help="Tag torrents with unregistered tracker status message",
    )

File: internal/commands/tagging.go
package commands

import (
	"fmt"
	"net/url"
	"strings"
	"time"

	"github.com/autobrr/go-qbittorrent"
	"github.com/buroa/qbtools/internal/config"
	"github.com/buroa/qbtools/internal/utils"
	"github.com/rs/zerolog/log"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	"golang.org/x/net/publicsuffix"
)

const (
	TrackerStatusUnregistered = "unregistered"
	TrackerStatusDown         = "down"
	TrackerStatusNotWorking   = "not-working"
)

var (
	unregisteredMatches = []string{
		"UNREGISTERED",
		"TORRENT NOT FOUND",
		"TORRENT IS NOT",
		"NOT REGISTERED",
		"NOT EXIST",
		"UNKNOWN TORRENT",
		"TRUMP",
		"RETITLED",
		"INFOHASH NOT FOUND",
		"TORRENT HAS BEEN DELETED",
		"DEAD",
		"DUPE",
		"COMPLETE SEASON UPLOADED",
		"PROBLEM",
		"POSTPONED",
		"SPECIFICALLY BANNED",
		"OTHER",
		"NUKED",
		"INVALID INFOHASH",
	}

	maintenanceMatches = []string{
		"DOWN",
		"UNREACHABLE",
		"BAD GATEWAY",
		"TRACKER UNAVAILABLE",
	}
)

// taggingOptions holds all the flags for the tagging command
type taggingOptions struct {
	excludeCategories []string
	excludeTags       []string
	addedOn           bool
	duplicates        bool
	expired           bool
	lastActivity      bool
	notLinked         bool
	notWorking        bool
	sites             bool
	trackerDown       bool
	unregistered      bool
}

// parseFlags extracts all command flags into a taggingOptions struct
func parseFlags(cmd *cobra.Command) (*taggingOptions, error) {
	opts := &taggingOptions{}
	var err error

	if opts.excludeCategories, err = cmd.Flags().GetStringSlice("exclude-category"); err != nil {
		return nil, err
	}
	if opts.excludeTags, err = cmd.Flags().GetStringSlice("exclude-tag"); err != nil {
		return nil, err
	}
	if opts.addedOn, err = cmd.Flags().GetBool("added-on"); err != nil {
		return nil, err
	}
	if opts.duplicates, err = cmd.Flags().GetBool("duplicates"); err != nil {
		return nil, err
	}
	if opts.expired, err = cmd.Flags().GetBool("expired"); err != nil {
		return nil, err
	}
	if opts.lastActivity, err = cmd.Flags().GetBool("last-activity"); err != nil {
		return nil, err
	}
	if opts.notLinked, err = cmd.Flags().GetBool("not-linked"); err != nil {
		return nil, err
	}
	if opts.notWorking, err = cmd.Flags().GetBool("not-working"); err != nil {
		return nil, err
	}
	if opts.sites, err = cmd.Flags().GetBool("sites"); err != nil {
		return nil, err
	}
	if opts.trackerDown, err = cmd.Flags().GetBool("tracker-down"); err != nil {
		return nil, err
	}
	if opts.unregistered, err = cmd.Flags().GetBool("unregistered"); err != nil {
		return nil, err
	}

	return opts, nil
}

// filterTorrents filters torrents based on exclude categories and tags
func filterTorrents(torrents []qbittorrent.Torrent, opts *taggingOptions) []qbittorrent.Torrent {
	filtered := torrents

	// Filter by categories
	if len(opts.excludeCategories) > 0 {
		var result []qbittorrent.Torrent
		for _, torrent := range filtered {
			excluded := false
			for _, exc := range opts.excludeCategories {
				if torrent.Category == exc {
					excluded = true
					break
				}
			}
			if !excluded {
				result = append(result, torrent)
			}
		}
		filtered = result
	}

	// Filter by tags
	if len(opts.excludeTags) > 0 {
		var result []qbittorrent.Torrent
		for _, torrent := range filtered {
			excluded := false
			for _, exc := range opts.excludeTags {
				if strings.Contains(torrent.Tags, exc) {
					excluded = true
					break
				}
			}
			if !excluded {
				result = append(result, torrent)
			}
		}
		filtered = result
	}

	return filtered
}

// getTrackerConfig returns the tracker configuration for a given torrent
func getTrackerConfig(torrent qbittorrent.Torrent, trackerMap map[string]config.TrackerConfig) *config.TrackerConfig {
	tracker := torrent.Tracker
	if tracker == "" {
		tracker = torrent.Trackers[0].Url
	}

	parsedURL, err := url.Parse(tracker)
	if err != nil {
		return nil
	}

	tldPlusOne, err := publicsuffix.EffectiveTLDPlusOne(parsedURL.Hostname())
	if err != nil {
		return nil
	}

	domain := strings.ToLower(tldPlusOne)
	if tc, exists := trackerMap[domain]; exists {
		return &tc
	}

	return nil
}

// checkTrackerMessages checks if any tracker messages contain the given matches
func checkTrackerMessages(messages []string, matches []string) bool {
	for _, match := range matches {
		for _, msg := range messages {
			if strings.Contains(msg, match) {
				return true
			}
		}
	}
	return false
}

// getTrackerStatusTags returns tracker status tags for a torrent
func getTrackerStatusTags(torrent qbittorrent.Torrent, opts *taggingOptions) []string {
	var messages []string

	for _, tracker := range torrent.Trackers {
		if tracker.Status == qbittorrent.TrackerStatusOK {
			return nil
		}
		messages = append(messages, strings.ToUpper(tracker.Message))
	}

	if opts.unregistered && checkTrackerMessages(messages, unregisteredMatches) {
		return []string{TrackerStatusUnregistered}
	}

	if opts.trackerDown && checkTrackerMessages(messages, maintenanceMatches) {
		return []string{TrackerStatusDown}
	}

	return []string{TrackerStatusNotWorking}
}

// NewTaggingCommand creates the tagging command
func NewTaggingCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "tagging",
		Short: "Tag torrents with various criteria",
		Long: `Tag torrents. This command can be used to tag torrents with various tags,
such as torrents that have not been active for a while, torrents that have not been
working for a while, torrents that have expired an ratio or seeding time, torrents
that have the same content path, etc.`,
		RunE: runTagging,
	}

	cmd.Flags().StringSlice("exclude-category", []string{}, "Exclude all torrents with these categories")
	cmd.Flags().StringSlice("exclude-tag", []string{}, "Exclude all torrents with these tags")
	cmd.Flags().Bool("added-on", false, "Tag torrents with added date (last 24h, 7 days, 30 days, etc)")
	cmd.Flags().Bool("duplicates", false, "Tag torrents with the same content path")
	cmd.Flags().Bool("expired", false, "Tag torrents that have an expired ratio or seeding time")
	cmd.Flags().Bool("last-activity", false, "Tag torrents with last activity date")
	cmd.Flags().Bool("not-linked", false, "Tag torrents with files without hardlinks or symlinks")
	cmd.Flags().Bool("not-working", false, "Tag torrents with not working tracker status")
	cmd.Flags().Bool("sites", false, "Tag torrents with site names")
	cmd.Flags().Bool("tracker-down", false, "Tag torrents with temporarily down trackers")
	cmd.Flags().Bool("unregistered", false, "Tag torrents with unregistered tracker status message")

	return cmd
}

func runTagging(cmd *cobra.Command, args []string) error {
	log.Info().Msg("Tagging torrents in qBittorrent...")

	client := qbittorrent.NewClient(qbittorrent.Config{
		Host:     viper.GetString("qbittorrent_host"),
		Username: viper.GetString("qbittorrent_username"),
		Password: viper.GetString("qbittorrent_password"),
	})

	if err := client.Login(); err != nil {
		return fmt.Errorf("failed to login to qBittorrent: %w", err)
	}

	ctx := cmd.Context()

	opts, err := parseFlags(cmd)
	if err != nil {
		return fmt.Errorf("failed to parse flags: %w", err)
	}

	// Get all torrents
	torrents, err := client.GetTorrents(qbittorrent.TorrentFilterOptions{
		IncludeTrackers: true,
	})
	if err != nil {
		return fmt.Errorf("failed to get torrents: %w", err)
	}

	// Filter torrents based on exclude options
	torrents = filterTorrents(torrents, opts)

	// Get tracker configuration
	var trackers []config.TrackerConfig
	if err := viper.UnmarshalKey("trackers", &trackers); err != nil {
		return fmt.Errorf("failed to unmarshal trackers config: %w", err)
	}

	// Build tracker URL mapping
	trackerMap := make(map[string]config.TrackerConfig)
	for _, tracker := range trackers {
		for _, url := range tracker.URLs {
			trackerMap[url] = tracker
		}
	}

	now := time.Now()
	paths := make(map[string]bool)

	for _, torrent := range torrents {
		var tagsToAdd []string
		trackerConfig := getTrackerConfig(torrent, trackerMap)

		// Date-based tags
		if opts.addedOn && torrent.AddedOn > 0 {
			tag := utils.CalculateDateTags("added", torrent.AddedOn, now)
			tagsToAdd = append(tagsToAdd, tag)
		}

		if opts.lastActivity && torrent.LastActivity > 0 {
			tag := utils.CalculateDateTags("activity", torrent.LastActivity, now)
			tagsToAdd = append(tagsToAdd, tag)
		}

		// Site tags
		if opts.sites {
			siteTag := "site:unmapped"
			if trackerConfig != nil {
				siteTag = fmt.Sprintf("site:%s", trackerConfig.Name)
			} else {
				log.Warn().Str("tracker", torrent.Tracker).Msg("No tracker config found for site tag")
			}
			tagsToAdd = append(tagsToAdd, siteTag)
		}

		// Tracker status tags
		if opts.unregistered || opts.trackerDown || opts.notWorking {
			tagsToAdd = append(tagsToAdd, getTrackerStatusTags(torrent, opts)...)
		}

		// Expired torrents
		if opts.expired && trackerConfig != nil {
			if (trackerConfig.Ratio != 0 && torrent.Ratio >= trackerConfig.Ratio) ||
				(trackerConfig.Days != 0 && torrent.SeedingTime >= utils.SecondsFromDays(trackerConfig.Days)) {
				tagsToAdd = append(tagsToAdd, "expired")
			}
		}

		// Duplicate content paths
		if opts.duplicates && torrent.ContentPath != "" {
			if paths[torrent.ContentPath] && torrent.ContentPath != torrent.SavePath {
				tagsToAdd = append(tagsToAdd, "dupe")
			} else {
				paths[torrent.ContentPath] = true
			}
		}

		// Not linked files
		if opts.notLinked && torrent.ContentPath != "" && !utils.IsLinked(torrent.ContentPath) {
			tagsToAdd = append(tagsToAdd, "not-linked")
		}

		// Split on comma and trim spaces
		currentTags := strings.Split(torrent.Tags, ",")
		for i := range currentTags {
			currentTags[i] = strings.TrimSpace(currentTags[i])
		}

		// Check if tagsToAdd are already present
		if !utils.AreSetsEqual(currentTags, tagsToAdd) {
			newTags := strings.Join(tagsToAdd, ", ")
			client.SetTags(ctx, []string{torrent.Hash}, newTags)
			log.Info().Str("tags", newTags).Str("name", torrent.Name).Str("hash", torrent.Hash).Msg("Updated torrent tags")
		}
	}

	log.Info().Msg("Finished tagging torrents in qBittorrent")
	return nil
}
